### ISSUE TITLE
Proper cache invalidation

### ISSUE THREAD

**Issue by untitaker** [untitaker/quickenv#6]
quickenv currently relies on the user to run `quickenv reload` to reload envrc.

I currently believe there is no way to fix this. envrcs are arbitrary code and don't provide information as to when they need to be re-executed.

one way to invalidate the cache is to do it when the envrc file changes. also, direnv _does_ have (poorly documented) stdlib functions to declare which other files need to be watched for changes. it's just that nobody appears to use those consistently, and there can be triggers for cache invalidation other than file changes.

a good example of this is sentry's envrc, which does system sanity checks of arbitary complexity, where there's no obvious set of files to watch for changes.

**Comment by untitaker** [comment]
here's some ideas:

* invalidate based on TTL: for example, just cache the env for one hour. that might be enough to make the terminal snappy again without risking really stale caches
* invalidate based on set of files
* invalidate based on checked out git commit (reset on git pull)
* invalidate based on shell PID/parent process PID, or rather, scope cache to current shell, terminal or tmux session
* or simply disable cache altogether (opt-in)

what you probably also want is an explicit subcommand to invalidate, but not recompute the cache. that command could be called from git pull/checkout hooks

probably need to search for some popular .envrcs via sourcegraph, though my feeling is that those are often used in corporate environments rather than OSS projects.

**Comment by danielcompton** [comment]
I don’t know where else to put this so I thought I’d write it here. Looking at https://github.com/getsentry/sentry/blob/master/.envrc, a lot of what’s happening doesn’t seem to need to run on every new shell. 

Not sure if you have the scope for this but I’d recommend separating things out into:

1. Things that have to happen once
2. Things that have to happen when code changes
3. Things that have to happen in a new shell

And only put number 3 in the envrc. You could look at something like https://github.com/github/scripts-to-rule-them-all as a pattern to handle 1 and 2.

**Comment by untitaker** [comment]
Indeed, most of the value proposition disappears if the envrc simply was faster. There are plans to do so but I don't expect that to be solved by tomorrow, and even then quickenvs cache is currently marginally faster than activating a virtualenv.

I wonder if quickenv should stop writing the cache if the envrc is fast, and cache it if it isn't. That's easy to figure out during execution. Just need to pick a threshold.
